rpc/src/connection.rs

```diff
@@ -0,0 +1,540 @@
+// Copyright 2022 labring. All rights reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+use common::request::{
+    RequestHeader, ResponseHeader, CLIENT_REQUEST_TIMEOUT, MAX_DATA_LENGTH, MAX_FILENAME_LENGTH,
+    MAX_METADATA_LENGTH, REQUEST_HEADER_SIZE, REQUEST_QUEUE_LENGTH, RESPONSE_HEADER_SIZE,
+};
+use log::{debug, error};
+use std::{
+    io::{Read, Write},
+    sync::{mpsc, Arc, Mutex, RwLock},
+    thread::sleep,
+    time::Duration,
+    vec,
+};
+use tokio::{
+    io::{AsyncReadExt, AsyncWriteExt},
+    net::tcp::{OwnedReadHalf, OwnedWriteHalf},
+};
+// use tokio::{
+//     io::{AsyncReadExt, AsyncWriteExt},
+//     net::TcpStream,
+// };
+
+enum ConnectionStatus {
+    Connected = 0,
+    Disconnected = 1,
+}
+
+pub struct ClientConnection {
+    pub server_address: String,
+    stream: Option<std::net::TcpStream>,
+    status: RwLock<ConnectionStatus>,
+    _send_lock: Arc<Mutex<()>>,
+}
+
+impl ClientConnection {
+    pub fn new(server_address: &str) -> Self {
+        Self {
+            server_address: server_address.to_string(),
+            stream: None,
+            status: RwLock::new(ConnectionStatus::Disconnected),
+            _send_lock: Arc::new(Mutex::new(())),
+        }
+    }
+
+    pub fn connect(&mut self) -> Result<(), Box<dyn std::error::Error>> {
+        let stream = std::net::TcpStream::connect(self.server_address.clone())?;
+        self.stream = Some(stream);
+        self.status = RwLock::new(ConnectionStatus::Connected);
```

Comment:
why not `*self.status.write().unwrap() = ConnectionStatus::Connected`

Reply:
Done